src/toast/tod/crosstalk.py

```diff
@@ -0,0 +1,473 @@
+# py37+
+# from __future__ import annotations
+
+"""Simulate crosstalk between detectors in ToD.
+
+A typical use of this is that in your pipeline script,
+
+1. Add `add_crosstalk_args(parser)` to your argparse parser to add cli args,
+2. Read & create `op_crosstalk` by adding
+
+
+    .. highlight:: python
+    .. code-block:: python
+
+        if args.crosstalk_matrix is not None:
+            op_crosstalk = OpCrosstalk.read(args)
+
+3. Apply the crosstalk matrix to your data after the ToD is prepared (e.g. signal & noise) by adding
+
+
+    .. highlight:: python
+    .. code-block:: python
+
+        if args.crosstalk_matrix is not None:
+            if comm.comm_world is not None:
+                comm.comm_world.barrier()
+            op_crosstalk.exec(data, "tot_signal")
+            if comm.comm_world is not None:
+                comm.comm_world.barrier()
+
+Note that you need to create your crosstalk matri(x|ces)
+in HDF5 container(s) beforehand. A simple class `SimpleCrosstalkMatrix`
+can assist you to write such file,
+with the simple conventions that
+it has the following datasets:
+
+1. names: ASCII names of your detectors, say of length ``n``
+2. data: ``n`` by ``n`` float64 array of your crosstalk matrix ``m``, where you expect
+
+    ToD_crosstalked = m @ ToD_original
+
+    (Assuming row-major as is standard in Python.)
+
+Note that each crosstalk matrix is dense,
+and when multiple matrices are supplied, they are effectively block-diagonal
+(where each given dense matrix is a block.)
+I.e. in order to provide a block-diagonal crosstalk matrix,
+each block (together with the names of the detectors in that block)
+should write to a seperate HDF5 file to avoid unnessary computation.
+"""
+
+import argparse
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import h5py
+import numpy as np
+
+try:
+    from numba import jit
```

Comment:
Just to note that numba is used only when available so it is an optional dep. And since the supported toast env at NERSC already have numba there, so it seems an easy gain to avoid creation of multiple Numpy objects which exerts unnecessary memory pressure.

Also, do you want to emit the warning message, which indirectly points people to install Numba, where they might not understand the subtlety in running that on hybrid MPI situation? May be change that to info or even debug level?

Reply:
Removed in commit a38d1d6